luals-addon/factorio/factorio-plugin/util.lua

```diff
@@ -170,6 +170,192 @@ local function add_chain_diff(chain_diff, diffs)
   end
 end
 
+---@alias SourceRange {from: integer, to: integer, length: integer, content: string}
+---@alias LexerState
+--- | '"code"'
+--- | '"short_string"'
+--- | '"long_string"'
+--- | '"short_comment"'
+--- | '"long_comment"'
+
+---Lexically analyze Lua source files for positions of strings and comments.
+---Notably, this needs to be able to handle 'long brackets', which are context-sensitive.
+---We should really only be doing this once per source file.
+---@param source string
+---@return SourceRange[]
+local function lex_lua_nonexecutables(source)
+  ---@type SourceRange[]
+  local ranges = {}
+  ---@type LexerState
+  local state = "code"
+  local cursor = 1 -- 1 is the first character in the source file.
+
+  local delimit = ""
+  local patterned_delimit = ""
+  local start = 0
+  local _end = 0
+
+  local char_escaped = false
+
+  ---check if the next character(s) are equal to the given string
+  ---@param query string
+  ---@return boolean
+  local function peek(query)
+    if cursor + #query > #source then return false end
+    return source:sub(cursor, cursor + #query - 1) == query
+  end
+
+  local function take(query)
+    if not peek(query) then return false end
+    cursor = cursor + #query
+    return true
+  end
+
+  ---Parse a opening long bracket, like `[[` or `[=[`
+  ---Assumes the first bracket has already been consumed.
+  ---@return boolean, integer | nil
+  local function parse_longbracket_open()
+    -- Consume all the '='s
+    local count = 0
+    while take("=") do
+      count = count + 1
+    end
+    if take("[") then
+      return true, count
+    else
+      return false, nil
+    end
+  end
+
+  local function append_range()
+    ranges[#ranges + 1] = {
+      from = start,
+      to = _end,
+      length = _end - start + 1,
+      content = source:sub(start, _end),
+    }
+  end
+
+  local start_clock = os.clock()
+
+  while cursor <= #source do
+    -- read this as a switch statement.
+    local origin = cursor
+
+    ---@type {[LexerState]: fun()}
+    local modes = {
```

Comment:
Maybe construct this table and `default` outside of loop, or outside the function scope entirely? Not sure if this gets optimized by `luac`.
Alternatively, use a `if`/`elseif`/`else` set instead?

Reply:
I actually don't know what exact version of Lua the Lua LS is using, but it seems to be some kind of Lua 5.4. Lua does not do the kind of optimization you mentioned here, so yes the table constructor would have to be moved out of the loop/function. Lua in general has a one pass compiler, so the only optimizations it does is some bare minimum constant folding for things like `1+ 1`. LuaJIT is a different story, but I'm pretty sure Lua LS does not use LuaJIT